File: dataModuleTester/src/main/java/com/aneonex/bitcoinchecker/tester/volley/generic/GzipVolleyRequest.java
package com.aneonex.bitcoinchecker.tester.volley.generic;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.util.HashMap;
import java.util.Map;
import java.util.zip.GZIPInputStream;

import com.android.volley.AuthFailureError;
import com.android.volley.Cache.Entry;
import com.android.volley.NetworkResponse;
import com.android.volley.ParseError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.Response.ErrorListener;
import com.android.volley.Response.Listener;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.HttpHeaderParser;
import com.aneonex.bitcoinchecker.tester.volley.CheckerErrorParsedError;
import com.aneonex.bitcoinchecker.tester.volley.UnknownVolleyError;

public abstract class GzipVolleyRequest<T> extends Request<T> {

	private final Listener<T> listener;
	private final ErrorListener errorListener;
	private final Map<String, String> headers;
	
	private RequestQueue requestQueue;
	private String redirectionUrl = null;
	private int redirectionCount;
	
	private Map<String, String> requestHeaders;
	private NetworkResponse networkResponse;
	private String responseString;
	
	private final static int MAX_REDIRECTION_COUNT = 3;
	
	public GzipVolleyRequest(String url, Listener<T> listener, ErrorListener errorListener) {
		super(Method.GET, url, errorListener);
		
		this.listener = listener;
		this.errorListener = errorListener;
		
		this.headers = new HashMap<String, String>();
		this.headers.put("Accept-Encoding", "gzip");
		this.headers.put("User-Agent", "Bitcoin Checker (gzip)");
	}
	
	public RequestQueue getRequestQueue() {
		return requestQueue;
	}
	
	@Override
	public String getUrl() {
		if(redirectionUrl!=null)
			return redirectionUrl;
		return super.getUrl();
	}
	
	@Override
    public Map<String, String> getHeaders() throws AuthFailureError {
		requestHeaders = headers!=null ? headers : super.getHeaders();
		return requestHeaders;
    }
	
	@Override
	public Request<?> setRequestQueue(RequestQueue requestQueue) {
		this.requestQueue = requestQueue;
		return super.setRequestQueue(requestQueue);
	}

	@Override
	public void deliverError(VolleyError error) {
		if(error!=null && error.networkResponse!=null) {
			final int statusCode = error.networkResponse.statusCode;
			if(statusCode==HttpURLConnection.HTTP_MOVED_PERM || statusCode==HttpURLConnection.HTTP_MOVED_TEMP) {
				String location = error.networkResponse.headers.get("Location");
				if(location!=null && redirectionCount<MAX_REDIRECTION_COUNT) {
					++redirectionCount;
					redirectionUrl = location;
					requestQueue.add(this);
					return;
				}
			}
		}
		if(errorListener instanceof ResponseErrorListener)
			((ResponseErrorListener)errorListener).onErrorResponse(getUrl(), requestHeaders, networkResponse, responseString, error);
		else
			super.deliverError(error);
	}
	
	@Override
	protected void deliverResponse(final T response) {
		if(listener instanceof ResponseListener)
			((ResponseListener<T>)listener).onResponse(getUrl(), requestHeaders, networkResponse, responseString, response);
		else
			listener.onResponse(response);
	}
	
	protected abstract T parseNetworkResponse(Map<String, String> headers, String responseString) throws Exception;
	
	@Override
	protected Response<T> parseNetworkResponse(NetworkResponse response) {
		try {
			networkResponse = response;
			String responseString = "";
			final String encoding = response.headers.get("Content-Encoding");
            if(encoding!=null && encoding.contains("gzip")) {
                responseString = decodeGZip(response.data);
            } else {
                responseString = new String(response.data, HttpHeaderParser.parseCharset(response.headers));
            }
            this.responseString = responseString;
            final Map<String, String> headers = response.headers;
            final Entry cacheHeaders = HttpHeaderParser.parseCacheHeaders(response);
            response = null;
			return Response.success(parseNetworkResponse(headers, responseString), cacheHeaders);
		} catch (CheckerErrorParsedError checkerErrorParsedError) {
			return Response.error(checkerErrorParsedError);
		} catch (Exception e) {
			return Response.error(new ParseError(e));
		} catch (Throwable e) {
			return Response.error(new UnknownVolleyError(e));
		}
	}
	
	private String decodeGZip(byte[] data) throws Exception {
        String responseString = "";

        ByteArrayInputStream bais = null;
        GZIPInputStream gzis = null;
        InputStreamReader reader = null;
        BufferedReader in = null;

        try {
            bais = new ByteArrayInputStream(data);
            gzis = new GZIPInputStream(bais);
            reader = new InputStreamReader(gzis);
            in = new BufferedReader(reader);

            String readed;
            while ((readed = in.readLine()) != null) {
                responseString += readed+"\n";
            }
        } catch (Exception e) {
            throw e;
        } finally {
            try {
                if(bais!=null)
                    bais.close();
                if(gzis!=null)
                    gzis.close();
                if(reader!=null)
                    reader.close();
                if(in!=null)
                    in.close();
            } catch (Exception e) {};
        }

        return responseString;
    }
}

File: dataModuleTester/src/main/java/com/aneonex/bitcoinchecker/tester/volley/generic/GzipVolleyRequest.kt
package com.aneonex.bitcoinchecker.tester.volley.generic

import com.android.volley.*
import com.android.volley.toolbox.HttpHeaderParser
import com.aneonex.bitcoinchecker.datamodule.model.PostRequestInfo
import com.aneonex.bitcoinchecker.tester.volley.CheckerErrorParsedError
import com.aneonex.bitcoinchecker.tester.volley.UnknownVolleyError
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.util.zip.GZIPInputStream

abstract class GzipVolleyRequest<T>(url: String?, private val postRequestInfo: PostRequestInfo?, private val listener: Response.Listener<T>, errorListener: Response.ErrorListener)
    : Request<T>(if (postRequestInfo == null) Method.GET else Method.POST, url, errorListener) {

    companion object {
        private const val MAX_REDIRECTION_COUNT = 3

        /** Default charset for JSON request.  */
        private val PROTOCOL_CHARSET = Charsets.UTF_8
    }

    private val initialErrorListener = errorListener
    private val headers: Map<String, String>
    var requestQueue: RequestQueue? = null
        private set
    private var redirectionUrl: String? = null
    private var redirectionCount = 0
    private var requestHeaders: Map<String, String>? = null
    private var networkResponse: NetworkResponse? = null
    private var responseString: String? = null

    init {
        if(method == Method.POST && postRequestInfo == null) throw IllegalArgumentException("Invalid arguments: postRequestInfo cannot be null for POST method")

        headers = mutableMapOf (
                "Accept-Encoding" to "gzip",
                "User-Agent" to "Bitcoin Checker (gzip)"
        )

        postRequestInfo?.headers?.let {
            for (item in it){
                headers[item.key] = item.value
            }
        }
    }

    override fun getUrl(): String {
        return if (redirectionUrl != null) redirectionUrl.toString() else super.getUrl()
    }

    override fun getBodyContentType(): String {
        return if(postRequestInfo != null) "application/json; charset=$PROTOCOL_CHARSET" else super.getBodyContentType()
    }

    override fun getBody(): ByteArray {
        return if(postRequestInfo != null) return postRequestInfo.body.toByteArray(PROTOCOL_CHARSET) else super.getBody()
    }

    @Throws(AuthFailureError::class)
    override fun getHeaders(): Map<String, String> {
        requestHeaders = headers //: super.getHeaders()
        return requestHeaders!!
    }

    override fun setRequestQueue(requestQueue: RequestQueue): Request<*> {
        this.requestQueue = requestQueue
        return super.setRequestQueue(requestQueue)
    }

    override fun deliverError(error: VolleyError) {
        if (error.networkResponse != null) {
            val statusCode = error.networkResponse.statusCode
            if (statusCode == HttpURLConnection.HTTP_MOVED_PERM || statusCode == HttpURLConnection.HTTP_MOVED_TEMP) {
                val location = error.networkResponse.headers["Location"]
                if (location != null && redirectionCount < MAX_REDIRECTION_COUNT) {
                    ++redirectionCount
                    redirectionUrl = location
                    requestQueue!!.add(this)
                    return
                }
            }
        }
        if (initialErrorListener is ResponseErrorListener)
            initialErrorListener.onErrorResponse(url, requestHeaders, networkResponse, responseString, error)
            else super.deliverError(error)
    }

    override fun deliverResponse(response: T) {
        if (listener is ResponseListener<*>) (listener as ResponseListener<T>).onResponse(url, requestHeaders, networkResponse, responseString, response) else listener.onResponse(response)
    }

    @Throws(Exception::class)
    protected abstract fun parseNetworkResponse(headers: Map<String?, String?>?, responseString: String?): T?

    override fun parseNetworkResponse(response: NetworkResponse): Response<T> {
//        var response: NetworkResponse? = response
        return try {
            networkResponse = response
            val encoding = response.headers["Content-Encoding"]
            val responseString = if (encoding != null && encoding.contains("gzip")) {
                decodeGZip(response.data)
            } else {
                String(response.data, charset(HttpHeaderParser.parseCharset(response.headers)))
            }
            this.responseString = responseString
            val headers = response.headers
            val cacheHeaders = HttpHeaderParser.parseCacheHeaders(response)
//            response = null
            Response.success(parseNetworkResponse(headers, responseString), cacheHeaders)
        } catch (checkerErrorParsedError: CheckerErrorParsedError) {
            Response.error(checkerErrorParsedError)
        } catch (e: Exception) {
            Response.error(ParseError(e))
        } catch (e: Throwable) {
            Response.error(UnknownVolleyError(e))
        }
    }

    @Throws(Exception::class)
    private fun decodeGZip(data: ByteArray): String {
        var responseString = ""
        var bais: ByteArrayInputStream? = null
        var gzis: GZIPInputStream? = null
        var reader: InputStreamReader? = null
        var bufReader: BufferedReader? = null
        try {
            bais = ByteArrayInputStream(data)
            gzis = GZIPInputStream(bais)
            reader = InputStreamReader(gzis)
            bufReader = BufferedReader(reader)
            var readed: String?
            while (bufReader.readLine().also { readed = it } != null) {
                responseString += """
                    $readed

                    """.trimIndent()
            }
        } catch (e: Exception) {
            throw e
        } finally {
            try {
                bais?.close()
                gzis?.close()
                reader?.close()
                bufReader?.close()
            } catch (e: Exception) {
            }
        }
        return responseString
    }
}